aip/general/0236.md

```diff
@@ -0,0 +1,321 @@
+To safely update resources, it is beneficial to test these changes via policy
+rollout APIs.
+
+Previews are a rollout safety mechanism for policy resources, which gives the
+customer the ability to validate the effect of their proposed changes against
+live production traffic prior to the changes going live.
+
+## Guidance
+
+### System changes
+
+As an API consumer, one can validate that the experiment introduces the intended
+changes. The endpoints described below activate these proposals. The
+experimental configurations propagate upstream to the data plane, but the
+existing live version of the resource continues to remain active. Importantly,
+while incoming requests are evaluated against experimental versions, only the
+live version affects active traffic. Concurrently, the system logs how the
+policy experiment proposals would affect traffic should they be promoted to
+be the live policy. This mechanism allows API consumer to observe changes
+proposed by an experiment without risking unintended side effects for live
+traffic.
+
+### Non-goals
+
+#### Non-policy resources
+
+This proposal is for a safety mechanism for policy rollouts only. Safe rollouts
+for resources that are not governed by "policies" are not in scope.
+
+A policy is a resource that provides rules that admit or deny access to other
+resources. Generally, the outcome of a policy can be evaluated to a specific set
+of outcomes. Firewall policies exemplify a case that is suitable for preview.
+The system simply logs the traffic that the firewall policy would accept or
+deny before making live changes to the configuration.
+
+### Nested collection called experiments
+
+A new configuration of a policy to be validated using preview mode is stored as
+a nested collection under that policy resource.
+
+A hypothetical policy Resource called, `Policy`, will be used throughout. It
+has the following resource name pattern:
+
+`projects/{project}/locations/{location}/policies/{policy}`
+
+The experimental versions of the resource used for preview mode are represented
+as a nested collection under `Policy`. The resource type will follow the
+convention "<resource_name> + Experiment”. The following is the pattern for the
+experiment:
+
+`projects/{project}/locations/{location}/policies/{policy}/experiments/{experiment}`
+
+A proto used to represent an experiment **must** contain the following fields:
+
+  1. The required top-level fields for a resource, like `name` and `etag`
+  2. The policy that is being tested itself
+  3. The field, `preview_metadata` for preview
+
+
+```proto
+message PolicyExperiment {
+
+  // The name of the PolicyExperiment.
+  string name = 1;
+ 
+  // The policy experiment. This Policy will be used to simulate the effects of
+  // the change but will affect life traffic.
+  Policy policy = 2;
+
+  // The metadata associated with this policy experiment.
+  PreviewMetadata preview_metadata = 3 [(google.api.field_behavior) = OUTPUT_ONLY];
```

Comment:
Unless we expect one common metadata type, I'd propose this has a name of *RegularResourceName*`ExperimentMetadata`. That means if there are multiple policy resources, they'd each have their own experiment resource type, each with its own metadata type.

Reply:
Yes, I also called this out. I'm unclear about what the resolution was on this topic. @nitinapi please explain?